### PR TITLE
Deterministic random ops accepting seed & offset

### LINKED ISSUE #535: Feature: allow explicit manipulation of Philox RNG state via functional version of RNG ops
The request is to implement
```cpp
TensorView functional_uniform(Val* philox_seed, Val* philox_offset, const std::vector<Val*>& shape,  Val* low,  Val* high, DataType dtype)
```
and the corresponding Python bindings. This operation should *not* increase nvFuser's `rand_offset` defined here https://github.com/NVIDIA/Fuser/blob/730215f36f679867ef25b56bd68dae17d13037f9/csrc/executor.cpp#L1460-L1468
because it's essentially a deterministic operation.

In addition, the result should be independent of the chosen vectorization, but I think it is already implemented by using linear index here
https://github.com/NVIDIA/Fuser/blob/730215f36f679867ef25b56bd68dae17d13037f9/csrc/codegen.cpp#L706

**Motivation**
Currently, nvFuser has two random ops `uniform` and `normal`. Internally it uses a Philox PRNG and its state is tied to PyTorch's RNG state. For each kernel invocation nvFuser is advancing PyTorch's RNG by the number of random ops in a kernel:
https://github.com/NVIDIA/Fuser/blob/730215f36f679867ef25b56bd68dae17d13037f9/csrc/executor_kernel_arg.cpp#L603-L614

That offers a great eager-like experience and `torch.manual_seed` and `torch.random.fork_rng` behaves as expected with nvFuser kernel invocations.

However, there are cases where explicit seed manipulation might be more efficient, for example, for dropout backward it's sufficient to save the Philox seed to recompute the dropout mask during backward instead of saving the mask from forward.

### PR DESCRIPTION
**Pull request by jacobhinkle**
This change introduces "functional" random operations which are deterministic versions of RNG ops that accept two additional `Val*` arguments for the seed and offset of the Philox RNG we use. Instead of calling `uniform(shape, low, high, dtype)` one can now call `uniform(shape, low, high, dtype, rng_seed=seed, rng_offset=offset)` which will generate deterministic code using `seed` and `offset` for the RNG. In the python frontend, these optional `rng_seed` and `rng_offset` are added to the `uniform` and `normal` ops. In the C++ backend, optional `philox_seed` and `philox_offset` arguments are added to the `uniform`, `normal`, `rand`, `randn`, `rand_like`, and `randn_like` ops.

Fixes #535.

### CODE REVIEW

File: csrc/ir/internal_nodes.h [hunk 1]
```diff
@@ -568,6 +568,79 @@ class TORCH_CUDA_CU_API RNGOp : public Expr {
   }
 };
 
+//! FunctionalRNGOp is just like RNGOp but takes explicit Vals for the Philox
+//! seed and offset so it is deterministic.
+class TORCH_CUDA_CU_API FunctionalRNGOp : public Expr {
```
Comment (zasdfgbnm) [inline csrc/ir/internal_nodes.h:573]: Do we need a new expr type? How is this behaving differently as the non-functional one? Using the same expr type is more aligned with the change I am planning to make as described in https://github.com/NVIDIA/Fuser/issues/535#issuecomment-1607849936, but if there are some blockers, I am also OK with adding this new expr type, but likely I will remove it in the future.

Comment (jacobhinkle) [reply csrc/ir/internal_nodes.h:573]: I started by trying to reuse `RNGOp`, but the seed and offset are additional inputs, and it was cumbersome to test whether the seed was specified each time it was used. But I think I now after going through the codegen part, I understand your approach. Creating that `GetPhiloxSeedAndOffsetFromATen` is much better since we could make use of that op for other expressions potentially. In this case, when determining whether to pass the additional `philox_args` argument, we would look in `summary_` for presence of that one `Expr`.

For this PR I could try making that change, but leaving the `kernel_inputs` part to you. Do you think that would make sense?

Comment (zasdfgbnm) [reply csrc/ir/internal_nodes.h:573]: That absolutely makes sense.

Comment (jacobhinkle) [reply csrc/ir/internal_nodes.h]: I just pushed a change that removes `FunctionalRNGOp`. Now `RNGOp` can take optional `Val*` arguments `philox_seed` and `philox_offset`. I also provide `RNGOp::getRNGSeedVal()` and `RNGOp::getRNGOffsetVal()` which return the `Val*`s that were provided (or `nullptr)` and `isDeterministic()` to test whether those are present. In codegen, I just check those pointers and use them instead of referring to `philox_args` if non-null.

File: csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -2327,11 +2329,19 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+            fd->recordingState(minval()),
+            fd->recordingState(maxval()),
+        };
+        if (rng_seed.has_value()) {
+          TORCH_CHECK(
+              rng_offset.has_value(),
+              "When providing rng_seed, rng_offset must also be provided");
+          arg_states.push_back(fd->recordingState(rng_seed.value()()));
+          arg_states.push_back(fd->recordingState(rng_offset.value()()));
+        }
```
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:2342]: @kevinstephano This is an instance of a pattern I hadn't seen yet. The `rng_seed` and `rng_offset` args are optional `Scalar`s. I am handling them by pushing them onto the back of `args_` since this automatically includes them in hashing and equality checks, but this feels kludgy since there are other possibly-positional args (`dtype` in this case) that come before `rng_seed` and `rng_offset`. I was thinking it would make sense here to make these kwonly args, but there is still the issue of how to handle them on the `RecordFunctor` side.

It seems like cases like this have been thought of to some degree, due to the presence of the `RecordFunctor::arg_names_` vector, which this PR provides a setter for. Currently I think that works alright in this case. It probably is always fine as long as classes first call `RecordFunctor::print()` then print their additional args only as kwargs.

**Comment by jacobhinkle** [comment]
!build

File: csrc/kernel_cache.cpp [hunk 3]
```diff
@@ -851,13 +851,21 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
       group_runtime_inputs.push(args_manager.checkTensorMap(input));
     }
 
-    // launch compileKernel thread here
-    getThreadPool()->run([=]() {
-      FUSER_PERF_SCOPE("FusionKernelRuntime::compileFusionParallel");
+    bool serial_compile = true;
+    if (serial_compile) {
```
Comment (zasdfgbnm) [inline csrc/kernel_cache.cpp:855]: Why the change here?

File: test/test_rng.cpp [hunk 4]
```diff
@@ -354,4 +354,67 @@ TEST_F(RNGTest, RandLikeReduction) {
   testValidate(fec.fusion(), {out}, {t0}, {t3}, __LINE__, __FILE__);
 }
 
+//! This is the same as the Uniform test, but we compare against
+//! functional_uniform in which we provide a seed and offset.
+TEST_F(RNGTest, FunctionalUniform) {
```
Comment (zasdfgbnm) [inline test/test_rng.cpp:359]: Could you also test the case where there is only functional uniform, no stochastic uniform in the fusion?

File: csrc/ir/nodes.cpp [hunk 5]
```diff
@@ -806,8 +808,15 @@ RNGOp::RNGOp(
   for (auto v : parameters) {
     addInput(v);
   }
+  if (philox_seed) {
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/ir/nodes.cpp:811]: Good catch. I have the proper check on both arguments in the python bindings but I will replicate that here.

File: csrc/kernel_cache.cpp (hunk 3 above)
Comment (jacobhinkle) [reply csrc/kernel_cache.cpp:855]: Ah. I forgot to reset this! I wind up doing this sometimes for easier debugging of compile failures. I'll fix this, and maybe add a small PR in the future that disables multithreaded compilation with an `NVFUSER_DISABLE` option.

File: test/test_rng.cpp (hunk 4 above)
Comment (jacobhinkle) [reply test/test_rng.cpp:359]: This test now also tests without any stochastic ops present.

**Comment by jacobhinkle** [comment]
!build

**Review by zasdfgbnm — COMMENTED** [review]
codegen and cpp test LGTM. Didn't check frontend. Will leave it to @kevinstephano to approve.

File: python_tests/test_python_frontend.py [hunk 6]
```diff
@@ -2359,6 +2359,49 @@ def fusion_func(fd: FusionDefinition) -> None:
...
+        param1 = 0.3  # low or mean
+        param2 = 1.7  # hi or sigma
+
+        for randopname in ["uniform", "normal"]:
+
+            def fusion_func(fd: FusionDefinition):
+                # Note: this is awkward, but we need to set the random seed
+                # each time the fusion is executed otherwise subsequent calls
+                # will compute t1 after advancing the seed
+                torch.manual_seed(0)
```
Comment (IvanYashchuk) [inline python_tests/test_python_frontend.py:2380]: This also might affect all other tests. We can use a special context manager to localize this seed manipulation:
```py
with torch.random.fork_rng(devices=[device.index]):
    torch.manual_seed(0)
    ...
```

File: python_tests/test_python_frontend.py
Comment (IvanYashchuk) [inline python_tests/test_python_frontend.py:1]: What do you think of adding a test that verifies the sequences generated by the stateful and stateless approaches match? Something along the lines of:
```py
from nvfuser import FusionDefinition, DataType
import torch
from random import randint

a = torch.ones(5, 9, device="cuda")

with FusionDefinition() as fd1:
    t1 = fd1.from_pytorch(a)
    min = fd1.define_scalar(0.0)
    max = fd1.define_scalar(1.0)
    shape = [fd1.define_scalar(5), fd1.define_scalar(9)]
    u = fd1.ops.uniform(min, max, shape)
    t2 = t1 * u
    fd1.add_output(t2)

with FusionDefinition() as fd2:
    t1 = fd2.from_pytorch(a)
    min = fd2.define_scalar(0.0)
    max = fd2.define_scalar(1.0)
    shape = [fd2.define_scalar(5), fd2.define_scalar(9)]
    rng_seed = fd2.define_scalar(DataType.Int)
    rng_offset = fd2.define_scalar(DataType.Int)
    u = fd2.ops.uniform(min, max, shape, rng_seed=rng_seed, rng_offset=rng_offset)
    t2 = t1 * u
    fd2.add_output(t2)

# Let's try three different random seeds
for _ in range(3):
    max_seed = 2**63-1
    seed = randint(0, max_seed)
    torch.manual_seed(seed)
    stateless_sequence = [fd2.execute([a, seed, rng_offset]) for rng_offset in range(0, 10*4, 4)]
    stateful_sequence = [fd1.execute([a]) for _ in range(10)]

    for i, (sful, sless) in enumerate(zip(stateful_sequence, stateless_sequence)):
        try:
            torch.testing.assert_close(sful[0], sless[0])
        except AssertionError as e:
            print(f"Assertion failed for iteration {i} with seed {seed}")
            print(e)
            break
```

File: csrc/kernel.cpp [hunk 7]
```diff
@@ -84,8 +84,15 @@ class KernelIrScanner : private IrVisitor {
...
 
   void handle(RNGOp* rng_op) final {
-    summary_.max_rng_offsets =
-        std::max<int>(summary_.max_rng_offsets, rng_op->getRNGOffset());
+    summary_.has_philox_op = true;
+    if (!rng_op->isDeterministic()) {
+      // NOTE: RNGOps that are provided a seed and offset should not contribute
+      // to max_rng_offsets, since that would cause the executor to increment
+      // the offset, and these types of deterministic RNGOp should not affect
+      // random ops at all.
```
Comment (IvanYashchuk) [inline csrc/kernel.cpp:92]: This is great!

File: csrc/python_frontend/python_bindings.cpp (hunk 2 above)
Comment (IvanYashchuk) [reply csrc/python_frontend/python_bindings.cpp:2380]: I fully support making `rng_seed` and `rng_offset` keyword-only arguments on the user-facing side.

**Review by IvanYashchuk — APPROVED** [review]
Looking great! For the same seed the results match with a sequence of stateful calls, I suggested including a test for this so that we won't accidentally break this property in the future.

File: python_tests/test_python_frontend.py
Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:1]: This is a great suggestion. Thanks! I have updated the test.

**Comment by zasdfgbnm** [comment]
Is there any reason why this is not merged?
